### PR TITLE
DOC-9596 document UI updates for udf js

### PR DESCRIPTION
**Pull request by RayOffiah**

### CODE REVIEW

File: modules/tools/pages/udfs-ui.adoc [hunk 1]
```diff
@@ -0,0 +1,70 @@
+= User-Defined Functions UI
+:description: The Couchbase Server admin console provides a UI for adding User-defined-functions
+
+[abstract]
+{description}
+
+
+Starting from Version 7.1, the Couchbase Administration console provides a CRUD{empty}footnote:[**CR**eate **U**pdate **D**elete] interface for user defined functions written in {sqlpp} or JavaScript.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: Doesn't the R in CRUD stand for "read"? 🤔 I would just rephrase this to avoid the acronym:

```suggestion
Starting from Couchbase Server 7.1, the Couchbase Administration console provides an interface to create, update, or delete user-defined functions written in {sqlpp} or JavaScript.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 2]
```diff
@@ -0,0 +1,70 @@
+= User-Defined Functions UI
+:description: The Couchbase Server admin console provides a UI for adding User-defined-functions
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc:2]: Consistent hyphenation — also capitalization, punctuation

```suggestion
:description: The Couchbase Server admin console provides a UI for adding user-defined functions.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+[abstract]
+{description}
+
+
+Starting from Version 7.1, the Couchbase Administration console provides a CRUD{empty}footnote:[**CR**eate **U**pdate **D**elete] interface for user defined functions written in {sqlpp} or JavaScript.
+
+Start at the administration console and select menu:Query[UDFs] from the menus to access the `UDF` screen:
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: We use bold text for UI elements. You can add the `[.ui]` role if you want, but it's not necessary, and it does make the source look more cluttered / harder to read.

Also, not actually a style rule, but I'd advise against using a colon at the end of a paragraph before an image. It's just one more thing to change if you ever move the paragraphs or images around, or add an extra paragraph.

```suggestion
Start at the administration console and select menu:Query[UDFs] from the menus to access the *UDF* screen.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+image::udfs-ui/udfs-screen.png[alt="screen for adding user-defined functions"]
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc:12]: Don't really need the subdirectory? All the other images for this module are in one flat directory

File: modules/tools/pages/udfs-ui.adoc [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+== Javascript Function Libraries
+
+Click on `{plus}{nbsp}add function library` link to create a library of JavaScript functions.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc:16]: UI element

```suggestion
Click on *{plus}{nbsp}add function library* link to create a library of JavaScript functions. 
```

File: modules/tools/pages/udfs-ui.adoc [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+== Javascript Function Libraries
+
+Click on `{plus}{nbsp}add function library` link to create a library of JavaScript functions. 
+
+image::udfs-ui/add-library-screen.png[,500, alt="add javascript library screen"]
+
+When you create a new library, an example function (`add`) is supplied as an example. 
+You can delete it when you add your own function, or leave it to serve as a basic example.
+
+You can set a `Namespace` for your library which will restrict its access to users that have permissions to access the bucket and scope defined in the Namespace.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: UI element

```suggestion
You can set a *Namespace* for your library which will restrict its access to users that have permissions to access the bucket and scope specified by the namespace.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+Before you leave the screen, you must set a `Library Name`.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: UI element

```suggestion
Before you leave the screen, you must set a *Library Name*.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+Press the btn:[Save] to leave the editor and return to the main UDF screen.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: I always like to include the **Cancel** button for completeness. Also, precision — the main purpose of the **Save** button is save the changes, not to leave the editor. 

```suggestion
Click btn:[Save] to save the library and return to the main UDF screen, or btn:[Cancel] to cancel without saving the library.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+image::udfs-ui/udf-screen-with-library-added.png[]
+
+You can `edit` the library again to add and/or delete functions contained inside it. 
+
+You can also delete the whole library by clicking on the `Drop` link.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: UI elements. Capitalization should match UI.

```suggestion
You can *edit* the library again to add and/or delete functions contained inside it. 

You can also delete the whole library by clicking on the *drop* link.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+[#user-defined-functions]
+== User-Defined Functions
+
+You can add your own user-defined functions in the lower `User-Defined Functions` list. These functions can be one of two types:
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: UI elements, semantic line breaks

```suggestion
You can add your own user-defined functions in the lower *User-Defined Functions* list.
These functions can be one of two types:
```

File: modules/tools/pages/udfs-ui.adoc [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+[horizontal]
+*Inline*:: This is a function written in {sqlpp} which can be used as part of another {sqlpp} statement (such as xref:n1ql:n1ql-language-reference/selectintro.adoc[`SELECT`] and xref:n1ql:n1ql-language-reference/execfunction.adoc[`EXECUTE FUNCTION`])
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: Actually tbh I would move this section down to the discussion of the **Function Type** drop-down list. It would be more useful there.

File: modules/tools/pages/udfs-ui.adoc [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+*JavaScript*:: You can create {sqlpp} functions that call the Javascript functions defined in your library.
+
+To add a user-defined function, click on `{plus}{nbsp}add function` below the `User-Defined Functions` list.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: UI elements

```suggestion
To add a user-defined function, click on *{plus}{nbsp}add function* below the *User-Defined Functions* list.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+image::udfs-ui/add-function.png[,500, alt="Add function screen"]
+
+The `Namespace` can be set to define the scope that function resides in (`travel-sample.inventory` for example).
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: UI elements.

```suggestion
The *Namespace* can be set to define the scope that function resides in (`travel-sample.inventory` for example). 
```

File: modules/tools/pages/udfs-ui.adoc [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+To add a user-defined function, click on `{plus}{nbsp}add function` below the `User-Defined Functions` list.
+
+image::udfs-ui/add-function.png[,500, alt="Add function screen"]
+
+The `Namespace` can be set to define the scope that function resides in (`travel-sample.inventory` for example). 
+
+Users will need permissions to access this namespace to run the function.
+
+Use any valid identifier as the `Function Name`.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: UI elements

```suggestion
Use any valid identifier as the *Function Name*.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+The name must be unique and is case-insensitive.
+
+You can define a list of fixed parameters for your function, or you can use the `...` symbol which indicates a variable length function list defined as `args[]`.
+
+As noted <<user-defined-functions, above>>, the `Function Type` can be either `inline` or `Javascript`.
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: UI elements. Also as noted above, I would move the description of Inline and External functions to this point.

```suggestion
As noted <<user-defined-functions, above>>, the *Function Type* can be either `inline` or `Javascript`. 
```

File: modules/tools/pages/udfs-ui.adoc [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+The name must be unique and is case-insensitive.
+
+You can define a list of fixed parameters for your function, or you can use the `...` symbol which indicates a variable length function list defined as `args[]`.
+
+As noted <<user-defined-functions, above>>, the `Function Type` can be either `inline` or `Javascript`. 
+
+If you are defining an `inline` function then you can fill in any valid {sqlpp} expression which can then be used as part of another {sqlpp} statement.
+This is the default when creating a function.
+
+If you switch `Function Type` to `Javascript` then the dialog will change to configure a JavaScript function call:
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc]: ui element. Also colon before images.

```suggestion
If you switch *Function Type* to `Javascript` then the dialog will change to configure a JavaScript function call.
```

File: modules/tools/pages/udfs-ui.adoc [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+image::udfs-ui/add-javascript-function.png[,500, alt="add Javascript function"]
+
+Then you select the `Javascript Library` where you created your Javascript function, and the name of your function. (There is no need to include the parameters).
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc:68]: ui element, semantic line breaks. Also can you move the information about the **Library Function Name** to its own paragraph.

```suggestion
Then you select the *Javascript Library* where you created your Javascript function, and the name of your function.
(There is no need to include the parameters.)
```

File: modules/tools/pages/udfs-ui.adoc [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+
```
Comment (simon-dew) [inline modules/tools/pages/udfs-ui.adoc:72]: Can you add a separate paragraphs explicitly describing:

1. The **Library Function Name** box.
2. The **Save Function** button and the **Cancel** button.

**Review by simon-dew — CHANGES_REQUESTED** [review]
Thanks Ray, this is great. Just one addition, could you also add a section about the UDF UI (including a link) to the [tools-ref](https://docs.couchbase.com/server/current/tools/tools-ref.html) page? Thanks.

File: modules/tools/pages/udfs-ui.adoc (hunk 1 above)
Comment (RayOffiah) [reply modules/tools/pages/udfs-ui.adoc]: Yup, stands for Read.
Mustn't write without coffee.

File: modules/tools/pages/udfs-ui.adoc (hunk 4 above)
Comment (RayOffiah) [reply modules/tools/pages/udfs-ui.adoc:12]: I think the images are better in their own folder that relates to the guide they're being used for. It makes them easier to find. 
I'm going to leave this one.
